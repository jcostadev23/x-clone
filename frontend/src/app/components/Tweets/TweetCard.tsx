import { Tweet } from "@/app/types";
import { formatDate } from "@/app/utils/dateFormater";
import Image from "next/image";
import xIcon from "../../../public/x-icon.jpg";
import BookMarkIcon from "../Icons/BookMarkIcon";
import LikeUnlikeButton from "../TweetButtons/LikeUnlikeButton";
import RepostIcon from "../Icons/RepostIcon";
import ShareIcon from "../Icons/ShareIcon";
import ViewIcon from "../Icons/ViewIcon";
import Comment from "../PostComment";

type Props = {
  tweet: Tweet;
};

const TweetCard: React.FC<Props> = ({ tweet }) => {
  return (
    <div className="flex flex-col gap-3 bg-gray-100 hover:bg-gray-200 p-3 border-solid border-2 border-gray-300 max-w-full">
      <div className="flex gap-3">
        <div>
          <Image
            className="rounded-xl"
            src={xIcon}
            alt={"X image"}
            width={26}
            height={26}
          />
        </div>
        <section>
          <div className="flex gap-2">
            <strong className="text-black text-md">{tweet.userName}</strong>
            <span
              className="text-sm"
              title={tweet?.date ? formatDate(tweet?.date) : "-"}
            >
              {tweet?.date ? formatDate(tweet?.date) : "-"}
            </span>
          </div>
          <h3 className="text-sm/7">{tweet.description}</h3>
        </section>
      </div>
      <div className="flex gap-2 m-3 justify-between max-w-full">
        <Comment tweet={tweet} />
        <div className="flex gap-1.5 items-center text-gray-600 hover:text-emerald-500">
          <RepostIcon />
          123k
        </div>
        <LikeUnlikeButton tweet={tweet} />
        <div className="flex gap-1.5 items-center text-gray-600 hover:text-blue-500">
          <ViewIcon /> 4.5m
        </div>
        <div className="flex gap-3 items-center">
          <BookMarkIcon />
          <ShareIcon />
        </div>
      </div>
    </div>
  );
};

export default TweetCard;
