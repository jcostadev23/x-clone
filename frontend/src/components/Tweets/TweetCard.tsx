import { Tweet } from "@/types";
import TweetCardFooter from "./TweetCardFooter";
import TweetCardHeader from "./TweetCardHeader";

type Props = {
  tweet: Tweet;
};

const TweetCard: React.FC<Props> = ({ tweet }) => {
  return (
    <div className="flex flex-col gap-3 bg-gray-100 hover:bg-gray-200 p-3 border-solid border-2 border-gray-300 max-w-full">
      <TweetCardHeader tweet={tweet} />
      <TweetCardFooter tweet={tweet} />
    </div>
  );
};

export default TweetCard;
