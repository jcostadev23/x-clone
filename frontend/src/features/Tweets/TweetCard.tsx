"use client";

import { Tweet } from "@/types";
import TweetCardFooter from "./TweetCardFooter";
import TweetCardHeader from "./components/TweetCardHeader";
import { useAppContext } from "@/hooks/useAppContext";
import { useReducerContext } from "@/hooks/useReducer";
import { useCallback, useState } from "react";
import { addComment, tweetLikeUnlike } from "@/utils/apiCalls";
import { fetchAndSetTweets } from "@/helpers";

type Props = {
  tweet: Tweet;
};

const TweetCard: React.FC<Props> = ({ tweet }) => {
  const { setIsLoading, user } = useAppContext();
  const { dispatch } = useReducerContext();
  const [formState, setFormState] = useState({
    isOpen: false,
    comment: "",
  });

  const liked = tweet.likes.includes(user?.userId as string);

  const handleLikeUnlikeClick = useCallback(async () => {
    if (!tweet._id || !user?.userId) {
      return false;
    }

    setIsLoading(true);
    const resp = await tweetLikeUnlike(tweet._id, user.userId);
    setIsLoading(false);

    if (!resp) {
      return false;
    }

    await fetchAndSetTweets(setIsLoading, dispatch);
  }, [dispatch, setIsLoading, tweet._id, user.userId]);

  const handleComment = useCallback(async () => {
    if (!formState.comment || !tweet._id) {
      return false;
    }

    setIsLoading(true);
    const response = await addComment(tweet._id, {
      id: tweet.userId,
      comment: formState.comment,
    });
    setIsLoading(false);

    if (!response) {
      return false;
    }

    await fetchAndSetTweets(setIsLoading, dispatch);
  }, [dispatch, formState.comment, setIsLoading, tweet._id, tweet.userId]);

  return (
    <div className="flex flex-col gap-3 bg-gray-100 hover:bg-gray-200 p-3 border-solid border-2 border-gray-300 max-w-full">
      <TweetCardHeader tweet={tweet} />
      <TweetCardFooter
        liked={liked}
        form={formState}
        setForm={setFormState}
        onSubmit={handleComment}
        likes={tweet.likes.length}
        onLike={handleLikeUnlikeClick}
        commentsAmount={tweet.comments.length}
      />
    </div>
  );
};

export default TweetCard;
