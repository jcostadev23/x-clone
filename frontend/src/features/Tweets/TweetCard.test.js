import { render } from "@testing-library/react";
import TweetCard from "./TweetCard";

jest.mock("@/hooks/useAppContext", () => ({
  useAppContext: () => ({
    setIsLoading: jest.fn(),
    user: { userId: "12345" },
  }),
}));

jest.mock("@/hooks/useReducer", () => ({
  useReducerContext: () => ({
    dispatch: jest.fn(),
  }),
}));

jest.mock("./Header", () => ({
  Header: {
    root: ({ children }) => <div data-testid="header-root">{children}</div>,
    avatar: () => <div data-testid="header-avatar" />,
    details: () => <div data-testid="header-details" />,
  },
}));

jest.mock("./Footer", () => {
  const Footer = jest.requireActual("./Footer");
  return { ...Footer };
});

describe("TweetCard snapshot", () => {
  it("renderiza corretamente", () => {
    const fakeTweet = {
      _id: "tweet1",
      userId: "user1",
      likes: ["user2"],
      comments: [{}, {}],
      description: "Um tweet de teste",
      date: "2025-08-10",
    };

    const { container } = render(<TweetCard tweet={fakeTweet} />);
    expect(container).toMatchSnapshot();
  });
});
