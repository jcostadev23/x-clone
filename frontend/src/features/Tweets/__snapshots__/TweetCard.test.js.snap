// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TweetCard snapshot renderiza corretamente 1`] = `
<div>
  <div
    class="flex flex-col gap-3 bg-gray-100 hover:bg-gray-200 p-3 border-solid border-2 border-gray-300 max-w-full"
  >
    <div
      data-testid="header-root"
    >
      <div
        data-testid="header-avatar"
      />
      <div
        data-testid="header-details"
      />
    </div>
    <div
      class="flex gap-2 m-3 justify-between max-w-full"
      data-cy="TweetCardFooter"
    >
      <div>
        <button
          class="flex gap-1.5 items-center text-gray-600 hover:text-blue-600"
          type="button"
        >
          <svg
            fill="currentColor"
            height="18"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 32C114.6 32 0 125.1 0 240c0 47.6 19.9 91.2 52.9 126.3C38 405.7 7 439.1 6.5 439.5c-6.6 7-8.4 17.2-4.6 26S14.4 480 24 480c61.5 0 110-25.7 139.1-46.3C192 442.8 223.2 448 256 448c141.4 0 256-93.1 256-208S397.4 32 256 32zm0 368c-26.7 0-53.1-4.1-78.4-12.1l-22.7-7.2-19.5 13.8c-14.3 10.1-33.9 21.4-57.5 29 7.3-12.1 14.4-25.7 19.9-40.2l10.6-28.1-20.6-21.8C69.7 314.1 48 282.2 48 240c0-88.2 93.3-160 208-160s208 71.8 208 160-93.3 160-208 160z"
            />
          </svg>
          2
        </button>
      </div>
      <div
        class="flex gap-1.5 items-center text-gray-600 hover:text-emerald-500"
      >
        <svg
          fill="currentColor"
          height="20"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M19 7a1 1 0 0 0-1-1h-8v2h7v5h-3l3.969 5L22 13h-3V7zM5 17a1 1 0 0 0 1 1h8v-2H7v-5h3L6 6l-4 5h3v6z"
          />
        </svg>
        123k
      </div>
      <button
        class="flex gap-1.5 items-center text-gray-600 hover:text-red-600"
        type="button"
      >
        <svg
          fill="currentColor"
          height="18"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
          />
        </svg>
        1
      </button>
      <div
        class="flex gap-1.5 items-center text-gray-600 hover:text-blue-500"
      >
        <svg
          fill="currentColor"
          height="18"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
          />
        </svg>
         4.5m
      </div>
      <div
        class="flex gap-3 items-center"
      >
        <svg
          class="text-gray-600 hover:text-blue-500"
          fill="currentColor"
          height="18"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 384 512"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M336 0H48C21.49 0 0 21.49 0 48v464l192-112 192 112V48c0-26.51-21.49-48-48-48zm0 428.43l-144-84-144 84V54a6 6 0 0 1 6-6h276c3.314 0 6 2.683 6 5.996V428.43z"
          />
        </svg>
        <svg
          class="text-gray-600 hover:text-blue-500"
          fill="currentColor"
          height="18"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="18"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2.58582L18.2071 8.79292L16.7929 10.2071L13 6.41424V16H11V6.41424L7.20711 10.2071L5.79289 8.79292L12 2.58582ZM3 18V14H5V18C5 18.5523 5.44772 19 6 19H18C18.5523 19 19 18.5523 19 18V14H21V18C21 19.6569 19.6569 21 18 21H6C4.34315 21 3 19.6569 3 18Z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
